src/middlewares/csrf_test.ts

```diff
@@ -0,0 +1,423 @@
+import { App } from "../app.ts";
+import type { FreshContext } from "../context.ts";
+import { csrf } from "./csrf.ts";
+import { beforeEach, describe, it } from "@std/testing/bdd";
+import { expect } from "@std/expect";
+import { spy } from "@std/testing/mock";
+
+const rawSimplePostHandler = async (ctx: FreshContext) => {
+  if (ctx.req.headers.get("content-type") === "application/json") {
+    return new Response((await ctx.req.json())["name"]);
+  } else {
+    return new Response((await ctx.req.formData()).get("name"));
+  }
+};
+
+describe("CSRF by Middleware", () => {
+  let simplePostHandler = spy(rawSimplePostHandler);
+
+  beforeEach(() => {
+    simplePostHandler = spy(rawSimplePostHandler);
+  });
+
+  describe("simple usage", () => {
+    const app = new App();
+
+    //
+    app.all("*", csrf());
+    app.get("/form", (_ctx: FreshContext) => new Response("<form></form>"));
+    app.post("/form", simplePostHandler);
+    app.put("/form", (_ctx: FreshContext) => new Response("OK"));
+    app.delete("/form", (_ctx: FreshContext) => new Response("OK"));
+    app.patch("/form", (_ctx: FreshContext) => new Response("OK"));
+    app.head("/form", (_ctx: FreshContext) => new Response("OK"));
+
+    const handler = app.handler();
+
+    describe("GET /form", () => {
+      it("should be 200 for any request", async () => {
+        const res = await handler(new Request("http://localhost/form"));
+
+        expect(res.status).toBe(200);
+        expect(await res.text()).toBe("<form></form>");
+      });
+    });
+
+    describe("HEAD /form", () => {
+      it("should be 200 for any request", async () => {
+        const res = await handler(
+          new Request("http://localhost/form", {
+            method: "HEAD",
+          }),
+        );
+
+        expect(res.status).toBe(200);
+      });
+    });
+
+    describe("POST /form", () => {
+      it("should be 200 for local request", async () => {
+        /*
+         * <form action="/form" method="POST"><input name="name" value="fresh" /></form>
+         * or
+         * <script>
+         * fetch('/form', {
+         *   method: 'POST',
+         *   headers: {
+         *     'content-type': 'application/x-www-form-urlencoded',
+         *   },
+         *   body: 'name=fresh',
+         * });
+         * </script>
+         */
```

Comment:
Nit: I'd delete these comments. I feel like they don't add any additional information that isn't already present a few lines later.